clippy_lints/src/dereference.rs

```diff
@@ -1362,7 +1362,12 @@ fn replace_types<'tcx>(
                     let projection = cx.tcx
                         .mk_projection(assoc_item.def_id, cx.tcx.mk_substs_trait(new_ty, []));
 
-                    if let Ok(projected_ty) = cx.tcx.try_normalize_erasing_regions(cx.param_env, projection)
+                    // Verify that `new_ty` implements the relevant trait before trying to normalize `projection` to
+                    // avoid hitting this `span_bug`:
+                    // https://github.com/rust-lang/rust/blob/695072daa6cc04045f2aa79d751d884ad5263080/compiler/rustc_trait_selection/src/traits/query/normalize.rs#L272-L275
+                    // See: https://github.com/rust-lang/rust/issues/107877
+                    if implements_trait(cx, new_ty, assoc_item.container_id(cx.tcx), List::empty())
```

Comment:
Shouldn't this pass the substitutions needed for the trait, rather than just an empty list?

Reply:
Yes, but do we have a way of knowing what those could could/should be?

I was under the impression we were making a best effort guess of "no parameters needed."

Maybe I'm misunderstanding, but I thought that's what was happening with `[]` here: https://github.com/rust-lang/rust-clippy/blob/a2c2b27348bc8c8212c1d3529ff1571bbd188bd6/clippy_lints/src/dereference.rs#L1362-L1363

Reply:
The parameters are stored in the projection predicate. You'd need to replace the self type but otherwise they should be correct. This is probably the reason for the ambiguity error in the first place. If both `Foo<X>` and `Foo<Y>` are implemented then `Foo<_>::Assoc` could refer to either.

Reply:
I think `implements_trait` returns false in this case which hides the error.

Reply:
It looks like your analysis was spot on.